network/p2p/unicast/manager.go

```diff
@@ -138,49 +244,108 @@ func (m *Manager) rawStreamWithProtocol(ctx context.Context,
 		if err != nil {
-
 			// if the connection was rejected due to invalid node id, skip the re-attempt
 			if strings.Contains(err.Error(), "failed to negotiate security protocol") {
-				return s, dialAddr, fmt.Errorf("invalid node id: %w", err)
+				return fmt.Errorf("failed to dial remote peer: %w", err)
 			}
 
 			// if the connection was rejected due to allowlisting, skip the re-attempt
 			if errors.Is(err, swarm.ErrGaterDisallowedConnection) {
-				return s, dialAddr, fmt.Errorf("target node is not on the approved list of nodes: %w", err)
+				return fmt.Errorf("target node is not on the approved list of nodes: %w", err)
 			}
 
-			errs = multierror.Append(errs, err)
-			continue
+			m.logger.Warn().
+				Err(err).
+				Str("peer_id", peerID.String()).
+				Int("attempt", dialAttempts).
+				Uint64("max_attempts", maxAttempts).
+				Msg("retrying peer dialing")
+
+			return retry.RetryableError(err)
+		}
+
+		return nil
+	}
+
+	// retryable func that will attempt to create the stream using the stream factory if connection exists
+	createStreamAttempts := 0
+	createStream := func(context.Context) error {
+		createStreamAttempts++
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context done before stream could be created (retry attempt: %d, errors: %w)", createStreamAttempts, errs)
+		default:
 		}
 
+		var err error
+		// add libp2p context value NoDial to prevent the underlying host from dialingComplete the peer while creating the stream
+		// we've already ensured that a connection already exists.
+		ctx = libp2pnet.WithNoDial(ctx, "application ensured connection to peer exists")
 		// creates stream using stream factory
 		s, err = m.streamFactory.NewStream(ctx, peerID, protocolID)
 		if err != nil {
 			// if the stream creation failed due to invalid protocol id, skip the re-attempt
 			if strings.Contains(err.Error(), "protocol not supported") {
-				return nil, dialAddr, fmt.Errorf("remote node is running on a different spork: %w, protocol attempted: %s", err, protocolID)
+				return fmt.Errorf("remote node is running on a different spork: %w, protocol attempted: %s", err, protocolID)
 			}
 			errs = multierror.Append(errs, err)
-			continue
+			return retry.RetryableError(errs)
 		}
+		return nil
+	}
 
-		break
+	isConnected, err := m.connStatus.IsConnected(peerID)
+	if err != nil {
+		return nil, dialAddr, err
 	}
 
-	if retries == maxAttempts {
-		return s, dialAddr, errs
+	// check connection status and attempt to dial the peer if dialing is not in progress
+	if !isConnected {
+		// return error if we can't start dialing
+		if m.dialingInProgress(peerID) {
+			return nil, dialAddr, NewDialInProgressErr(peerID)
+		}
+		defer m.dialingComplete(peerID)
+
+		start := time.Now()
+		err = retry.Do(ctx, backoff, dialPeer)
+		duration := time.Since(start)
+		if err != nil {
+			if uint64(dialAttempts) == maxAttempts {
+				err = fmt.Errorf("failed to dial peer max attempts reached %d: %w", maxAttempts, err)
+			}
```

Comment:
I think this deserves a dedicated sentinel to indicate to the caller that the max attempts have been exhausted. To me, it would make the most sense, if `dialPeer` contained the logic to generate this specific sentinel. using in-line variables to extract information from `dialPeer` is a pattern that we should use only when absolutely necessary.

Reply:
https://github.com/onflow/flow-go/pull/3902/commits/6fd3ae44d2ed486454d844b0fa5e9491bf87db60